include/zephyr/drivers/i3c.h

```diff
@@ -0,0 +1,1596 @@
+
+	/** No error (not official error code) */
+	I3C_ERROR_TE_NONE,
+
+	I3C_ERROR_TE_MAX = I3C_ERROR_TE_UNKNOWN,
+	I3C_ERROR_TE_INVALID,
+};
+
+/*
+ * I3C_MSG_* are I3C Message flags.
+ */
+
+/** Write message to I3C bus. */
+#define I3C_MSG_WRITE			(0U << 0U)
+
+/** Read message from I2C bus. */
+#define I3C_MSG_READ			BIT(0)
+
+/** @cond INTERNAL_HIDDEN */
+#define I3C_MSG_RW_MASK			BIT(0)
+/** @endcond  */
+
+/** Send STOP after this message. */
+#define I3C_MSG_STOP			BIT(1)
+
+/** RESTART I3C transaction for this message.
+ *
+ * @note Not all I3C drivers have or require explicit support for this
+ * feature. Some drivers require this be present on a read message
+ * that follows a write, or vice-versa.  Some drivers will merge
+ * adjacent fragments into a single transaction using this flag; some
+ * will not.
+ */
+#define I3C_MSG_RESTART			BIT(2)
+
+/** Transfer use HDR mode */
+#define I3C_MSG_HDR			BIT(3)
+
+/** I3C HDR Mode 0 */
+#define I3C_MSG_HDR_MODE0		BIT(0)
+
+/** I3C HDR Mode 1 */
+#define I3C_MSG_HDR_MODE1		BIT(1)
+
+/** I3C HDR Mode 2 */
+#define I3C_MSG_HDR_MODE2		BIT(2)
+
+/** I3C HDR Mode 3 */
+#define I3C_MSG_HDR_MODE3		BIT(3)
+
+/** I3C HDR Mode 4 */
+#define I3C_MSG_HDR_MODE4		BIT(4)
+
+/** I3C HDR Mode 5 */
+#define I3C_MSG_HDR_MODE5		BIT(5)
+
+/** I3C HDR Mode 6 */
+#define I3C_MSG_HDR_MODE6		BIT(6)
+
+/** I3C HDR Mode 7 */
+#define I3C_MSG_HDR_MODE7		BIT(7)
+
+/** I3C HDR-DDR (Double Data Rate) */
+#define I3C_MSG_HDR_DDR			I3C_MSG_HDR_MODE0
+
+/** I3C HDR-TSP (Ternary Symbol Pure-bus) */
+#define I3C_MSG_HDR_TSP			I3C_MSG_HDR_MODE1
+
+/** I3C HDR-TSL (Ternary Symbol Legacy-inclusive-bus) */
+#define I3C_MSG_HDR_TSL			I3C_MSG_HDR_MODE2
+
+/** I3C HDR-BT (Bulk Transport) */
+#define I3C_MSG_HDR_BT			I3C_MSG_HDR_MODE3
+
+/**
+ * @brief One I3C Message.
+ *
+ * This defines one I3C message to transact on the I3C bus.
+ *
+ * @note Some of the configurations supported by this API may not be
+ * supported by specific SoC I3C hardware implementations, in
+ * particular features related to bus transactions intended to read or
+ * write data from different buffers within a single transaction.
+ * Invocations of i3c_transfer() may not indicate an error when an
+ * unsupported configuration is encountered.  In some cases drivers
+ * will generate separate transactions for each message fragment, with
+ * or without presence of @ref I3C_MSG_RESTART in #flags.
+ */
+struct i3c_msg {
+	/** Data buffer in bytes */
+	uint8_t			*buf;
+
+	/** Length of buffer in bytes */
+	uint32_t		len;
+
+	/** Flags for this message */
+	uint8_t			flags;
+
+	/**
+	 * HDR mode (@c I3C_MSG_HDR_MODE*) for transfer
+	 * if @c I3C_MSG_HDR is set in @c flags.
+	 */
+	uint8_t			hdr_mode;
+};
+
+/**
+ * @cond INTERNAL_HIDDEN
+ *
+ * These are for internal use only, so skip these in
+ * public documentation.
+ */
+struct i3c_target_desc;
+struct i3c_i2c_device_desc;
+
+__subsystem struct i3c_driver_api {
+	int (*do_daa)(const struct device *dev);
```

Comment:
There is no requirement saying that these must use typedef. My reason for not using typedef is simply that, under IDE, you can see all the function signatures inside the struct in code reference pop-ups, instead of needing to do one redirect to see the signature.